How to install dependency on Mac?

I've installed gsl, but looks like it wasn't enough:
```
>$ brew install gsl
==> Downloading https://homebrew.bintray.com/bottles/gsl-2.2.1.sierra.bottle.tar.gz
######################################################################## 100.0%
==> Pouring gsl-2.2.1.sierra.bottle.tar.gz
🍺  /usr/local/Cellar/gsl/2.2.1: 272 files, 8.6M
```
Here's the error when I try to build:
```
>$ cargo build
   Compiling GSL v0.4.28
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-m64" "-L" "/usr/local/lib/rustlib/x86_64-apple-darwin/lib" "/Users/franciswang/test/target/debug/deps/rgsl-6e9ac6569d203bcd.0.o" "-o" "/Users/franciswang/test/target/debug/deps/librgsl-6e9ac6569d203bcd.dylib" "/Users/franciswang/test/target/debug/deps/rgsl-6e9ac6569d203bcd.metadata.o" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/Users/franciswang/test/target/debug/deps" "-L" "/usr/local/lib/rustlib/x86_64-apple-darwin/lib" "-l" "gsl" "-l" "gslcblas" "-Wl,-force_load,/var/folders/x_/c899j8zj1p9617nktzy_55k00000gp/T/rustc.l05wzdpO5EaY/liblibc-6ec63c5a0e74a074.rlib" "-L" "/Users/franciswang/test/target/debug/deps" "-l" "c_vec-404a8700dec5f2e1" "-L" "/usr/local/lib/rustlib/x86_64-apple-darwin/lib" "-l" "std-a4729905" "/var/folders/x_/c899j8zj1p9617nktzy_55k00000gp/T/rustc.l05wzdpO5EaY/libcompiler_builtins-a4729905.rlib" "-l" "System" "-l" "pthread" "-l" "c" "-l" "m" "-dynamiclib" "-Wl,-dylib"
  = note: Undefined symbols for architecture x86_64:
  "_gsl_sf_legendre_array_size", referenced from:
      rgsl::legendre::associated_polynomials::legendre_array_size::hba1a0292657ad723 in rgsl-6e9ac6569d203bcd.0.o
  "_gsl_sf_legendre_sphPlm_deriv_array", referenced from:
      rgsl::legendre::associated_polynomials::legendre_sphPlm_deriv_array::h5e62b502224dfcf2 in rgsl-6e9ac6569d203bcd.0.o
  "_gsl_sf_legendre_sphPlm_array", referenced from:
      rgsl::legendre::associated_polynomials::legendre_sphPlm_array::hca6ca3693272788e in rgsl-6e9ac6569d203bcd.0.o
  "_gsl_bspline_deriv_alloc", referenced from:
      rgsl::types::basis_spline::BSpLineDerivWorkspace::new::h15a8433a124eadb4 in rgsl-6e9ac6569d203bcd.0.o
  "_gsl_bspline_deriv_free", referenced from:
      _$LT$rgsl..types..basis_spline..BSpLineDerivWorkspace$u20$as$u20$core..ops..Drop$GT$::drop::he88b99acc9916c98 in rgsl-6e9ac6569d203bcd.0.o
  "_gsl_sf_legendre_Plm_array", referenced from:
      rgsl::legendre::associated_polynomials::legendre_Plm_array::h61d8ad280cb37285 in rgsl-6e9ac6569d203bcd.0.o
  "_gsl_sf_legendre_Plm_deriv_array", referenced from:
      rgsl::legendre::associated_polynomials::legendre_Plm_deriv_array::hb8b0d48c1dd07921 in rgsl-6e9ac6569d203bcd.0.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)


error: aborting due to previous error

error: Could not compile `GSL`.

To learn more, run the command again with --verbose.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,18 +1,31 @@
 language: rust
-rust:
-- nightly
-- stable
+matrix:
+  include:
+    - os: linux
+      rust: stable
+      dist: trusty
+    - os: linux
+      rust: nightly
+      dist: trusty
+    - os: osx
+      rust: stable
+    - os: osx
+      rust: nightly
 sudo: true
 env:
   global:
     - LD_LIBRARY_PATH=/usr/local/lib
-install:
-  - sudo apt-get -qq update
-  - sudo apt-get install libgsl0-dev
+addons:
+  apt:
+    packages:
+    - libgsl0-dev
+before_install:
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update; fi
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install gsl; fi
 script:
   - rustc --version
-  - RUST_BACKTRACE=1 cargo build
-  - RUST_BACKTRACE=1 cargo doc
-  - RUST_BACKTRACE=1 cargo test
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then cargo build --features v2; else cargo build; fi
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then cargo doc --features v2; else cargo doc; fi
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then cargo test --features v2; else cargo test; fi
   - cd examples
-  - cargo build
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then cargo build --features v2; else cargo build; fi

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -18,3 +18,6 @@ libc = "~0.2"
 [lib]
 name = "rgsl"
 crate-type = ["dylib", "rlib"]
+
+[features]
+v2 = []

--- a/README.md
+++ b/README.md
@@ -26,6 +26,16 @@ Add the following line to your source code:
 extern crate rgsl;
 ```
 
+##Building
+
+Just run `cargo build`. However, if you have issues with GSL version 2.x, try building it with:
+
+```bash
+cargo build --features v2
+```
+
+If a project depends on this version, don't forget to add in your `Cargo.toml`!
+
 ##Documentation
 
 You can access the __rgsl__ documentation locally, just build it:

--- a/examples/Cargo.toml
+++ b/examples/Cargo.toml
@@ -3,6 +3,9 @@ name = "gsl-examples"
 version = "0.4.6"
 authors = ["Guillaume Gomez <guillaume1.gomez@gmail.com>"]
 
+[features]
+v2 = ["GSL/v2"]
+
 [dependencies.GSL]
 path = "../"
 

--- a/src/ffi.rs
+++ b/src/ffi.rs
@@ -536,14 +536,19 @@ extern "C" {
     // Associated Legendre Polynomials and Spherical Harmonics
     pub fn gsl_sf_legendre_Plm(l: c_int, m: c_int, x: c_double) -> c_double;
     pub fn gsl_sf_legendre_Plm_e(l: c_int, m: c_int, x: c_double, result: *mut gsl_sf_result) -> enums::Value;
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_sf_legendre_Plm_array(lmax: c_int, m: c_int, x: c_double, result_array: *mut c_double) -> enums::Value;
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_sf_legendre_Plm_deriv_array(lmax: c_int, m: c_int, x: c_double, result_array: *mut c_double,
         result_deriv_array: *mut c_double) -> enums::Value;
     pub fn gsl_sf_legendre_sphPlm(l: c_int, m: c_int, x: c_double) -> c_double;
     pub fn gsl_sf_legendre_sphPlm_e(l: c_int, m: c_int, x: c_double, result: *mut gsl_sf_result) -> enums::Value;
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_sf_legendre_sphPlm_array(lmax: c_int, m: c_int, x: c_double, result_array: *mut c_double) -> enums::Value;
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_sf_legendre_sphPlm_deriv_array(lmax: c_int, m: c_int, x: c_double, result_array: *mut c_double,
         result_deriv_array: *mut c_double) -> enums::Value;
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_sf_legendre_array_size(lmax: c_int, m: c_int) -> enums::Value;
     // Conical functions
     pub fn gsl_sf_conicalP_half(lambda: c_double, x: c_double) -> c_double;
@@ -1358,7 +1363,9 @@ extern "C" {
     // Basis Splines
     pub fn gsl_bspline_alloc(k: size_t, nbreak: size_t) -> *mut gsl_bspline_workspace;
     pub fn gsl_bspline_free(w: *mut gsl_bspline_workspace);
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_bspline_deriv_alloc(k: size_t) -> *mut gsl_bspline_deriv_workspace;
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_bspline_deriv_free(w: *mut gsl_bspline_deriv_workspace);
     pub fn gsl_bspline_knots(breakpts: *mut gsl_vector, w: *mut gsl_bspline_workspace) -> enums::Value;
     pub fn gsl_bspline_knots_uniform(a: c_double, b: c_double, w: *mut gsl_bspline_workspace) -> enums::Value;
@@ -1366,8 +1373,10 @@ extern "C" {
     pub fn gsl_bspline_eval_nonzero(x: c_double, Bk: *mut gsl_vector, istart: *mut size_t, iend: *mut size_t,
         w: *mut gsl_bspline_workspace) -> enums::Value;
     pub fn gsl_bspline_ncoeffs(w: *mut gsl_bspline_workspace) -> size_t;
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_bspline_deriv_eval(x: c_double, nderiv: size_t, dB: *mut gsl_matrix, w: *mut gsl_bspline_workspace,
         dw: *mut gsl_bspline_deriv_workspace) -> enums::Value;
+    #[cfg(not(feature = "v2"))]
     pub fn gsl_bspline_deriv_eval_nonzero(x: c_double, nderiv: size_t, Bk: *mut gsl_matrix, istart: *mut size_t, iend: *mut size_t,
         w: *mut gsl_bspline_workspace, dw: *mut gsl_bspline_deriv_workspace) -> enums::Value;
     pub fn gsl_bspline_greville_abscissa(i: size_t, w: *mut gsl_bspline_workspace) -> c_double;
@@ -2693,6 +2702,7 @@ pub struct gsl_bspline_workspace {
 }
 
 #[repr(C)]
+#[cfg(not(feature = "v2"))]
 pub struct gsl_bspline_deriv_workspace {
     pub k: size_t,          // spline order
     pub A: *mut gsl_matrix, // work matrix

--- a/src/legendre.rs
+++ b/src/legendre.rs
@@ -137,11 +137,13 @@ pub mod associated_polynomials {
     }
 
     /// This function computes arrays of Legendre polynomials P_l^m(x) and derivatives dP_l^m(x)/dx, for m >= 0, l = |m|, ..., lmax, |x| <= 1.
+    #[cfg(not(feature = "v2"))]
     pub fn legendre_Plm_array(m: i32, x: f64, result_array: &mut [f64]) -> enums::Value {
         unsafe { ffi::gsl_sf_legendre_Plm_array(result_array.len() as i32, m, x, result_array.as_mut_ptr()) }
     }
 
     /// This function computes arrays of Legendre polynomials P_l^m(x) and derivatives dP_l^m(x)/dx, for m >= 0, l = |m|, ..., lmax, |x| <= 1.
+    #[cfg(not(feature = "v2"))]
     pub fn legendre_Plm_deriv_array(m: i32, x: f64, result_array: &mut [f64], result_deriv_array: &mut [f64]) -> enums::Value {
         unsafe { ffi::gsl_sf_legendre_Plm_deriv_array(result_array.len() as i32, m, x, result_array.as_mut_ptr(),
             result_deriv_array.as_mut_ptr()) }
@@ -165,17 +167,20 @@ pub mod associated_polynomials {
     }
 
     /// This function computes arrays of normalized associated Legendre functions \sqrt{(2l+1)/(4\pi)} \sqrt{(l-m)!/(l+m)!} P_l^m(x), and derivatives, for m >= 0, l = |m|, ..., lmax, |x| <= 1.0
+    #[cfg(not(feature = "v2"))]
     pub fn legendre_sphPlm_array(m: i32, x: f64, result_array: &mut [f64]) -> enums::Value {
         unsafe { ffi::gsl_sf_legendre_sphPlm_array(result_array.len() as i32, m, x, result_array.as_mut_ptr()) }
     }
 
     /// This function computes arrays of normalized associated Legendre functions \sqrt{(2l+1)/(4\pi)} \sqrt{(l-m)!/(l+m)!} P_l^m(x), and derivatives, for m >= 0, l = |m|, ..., lmax, |x| <= 1.0
+    #[cfg(not(feature = "v2"))]
     pub fn legendre_sphPlm_deriv_array(m: i32, x: f64, result_array: &mut [f64], result_deriv_array: &mut [f64]) -> enums::Value {
         unsafe { ffi::gsl_sf_legendre_sphPlm_deriv_array(result_array.len() as i32, m, x, result_array.as_mut_ptr(),
             result_deriv_array.as_mut_ptr()) }
     }
 
     /// This function returns the size of result_array[] needed for the array versions of P_l^m(x), lmax - m + 1.
+    #[cfg(not(feature = "v2"))]
     pub fn legendre_array_size(lmax: i32, m: i32) -> enums::Value {
         unsafe { ffi::gsl_sf_legendre_array_size(lmax, m) }
     }

--- a/src/macros.rs
+++ b/src/macros.rs
@@ -18,3 +18,10 @@ macro_rules! rgsl_error(
         }
     );
 );
+
+#[macro_export]
+macro_rules! ffi_wrap {
+    ($name:tt, $cast:tt) => {
+        unsafe { ffi::FFI::wrap(ffi::$name as *mut ffi::$cast) }
+    }
+}

--- a/src/types/basis_spline.rs
+++ b/src/types/basis_spline.rs
@@ -47,7 +47,10 @@ A large collection of B-spline routines is available in the PPPACK library avail
 http://www.netlib.org/pppack, which is also part of SLATEC.
 !*/
 
+#[cfg(not(feature = "v2"))]
 use types::{VectorF64, MatrixF64};
+#[cfg(feature = "v2")]
+use types::VectorF64;
 use ffi;
 use enums;
 
@@ -152,10 +155,12 @@ impl ffi::FFI<ffi::gsl_bspline_workspace> for BSpLineWorkspace {
     }
 }
 
+#[cfg(not(feature = "v2"))]
 pub struct BSpLineDerivWorkspace {
     w: *mut ffi::gsl_bspline_deriv_workspace
 }
 
+#[cfg(not(feature = "v2"))]
 impl BSpLineDerivWorkspace {
     /// This function allocates a workspace for computing the derivatives of a B-spline basis
     /// function of order k.
@@ -206,6 +211,7 @@ impl BSpLineDerivWorkspace {
     }
 }
 
+#[cfg(not(feature = "v2"))]
 impl Drop for BSpLineDerivWorkspace {
     fn drop(&mut self) {
         unsafe { ffi::gsl_bspline_deriv_free(self.w) };
@@ -213,6 +219,7 @@ impl Drop for BSpLineDerivWorkspace {
     }
 }
 
+#[cfg(not(feature = "v2"))]
 impl ffi::FFI<ffi::gsl_bspline_deriv_workspace> for BSpLineDerivWorkspace {
     fn wrap(r: *mut ffi::gsl_bspline_deriv_workspace) -> BSpLineDerivWorkspace {
         BSpLineDerivWorkspace {

--- a/src/types/eigen_symmetric_workspace.rs
+++ b/src/types/eigen_symmetric_workspace.rs
@@ -804,7 +804,7 @@ fn eigen_symmetric_workspace() {
     use MatrixF64;
     use VectorF64;
 
-    let e = EigenSymmetricWorkspace::new(3).unwrap();
+    let e = EigenSymmetricWorkspace::new(2).unwrap();
     let m = MatrixF64::new(2, 2).unwrap();
 
     let data = [5., 5., 1., 6.];

--- a/src/types/interpolation.rs
+++ b/src/types/interpolation.rs
@@ -163,20 +163,20 @@ impl InterpType {
 
     /// Linear interpolation. This interpolation method does not require any additional memory.
     pub fn linear() -> InterpType {
-        ffi::FFI::wrap(ffi::gsl_interp_linear as *mut ffi::gsl_interp_type)
+        ffi_wrap!(gsl_interp_linear, gsl_interp_type)
     }
 
     /// Polynomial interpolation. This method should only be used for interpolating small numbers of points because polynomial interpolation
     /// introduces large oscillations, even for well-behaved datasets. The number of terms in the interpolating polynomial is equal to the
     /// number of points.
     pub fn polynomial() -> InterpType {
-        ffi::FFI::wrap(ffi::gsl_interp_polynomial as *mut ffi::gsl_interp_type)
+        ffi_wrap!(gsl_interp_polynomial, gsl_interp_type)
     }
 
     /// Cubic spline with natural boundary conditions. The resulting curve is piecewise cubic on each interval, with matching first and second
     /// derivatives at the supplied data-points. The second derivative is chosen to be zero at the first point and last point.
     pub fn cspline() -> InterpType {
-        ffi::FFI::wrap(ffi::gsl_interp_cspline as *mut ffi::gsl_interp_type)
+        ffi_wrap!(gsl_interp_cspline, gsl_interp_type)
     }
 
     /// Cubic spline with periodic boundary conditions. The resulting curve is piecewise cubic on each interval, with matching first and second
@@ -184,17 +184,17 @@ impl InterpType {
     /// the data must have the same y-value as the first point, otherwise the resulting periodic interpolation will have a discontinuity at
     /// the boundary.
     pub fn cspline_periodic() -> InterpType {
-        ffi::FFI::wrap(ffi::gsl_interp_cspline_periodic as *mut ffi::gsl_interp_type)
+        ffi_wrap!(gsl_interp_cspline_periodic, gsl_interp_type)
     }
 
     /// Non-rounded Akima spline with natural boundary conditions. This method uses the non-rounded corner algorithm of Wodicka.
     pub fn akima() -> InterpType {
-        ffi::FFI::wrap(ffi::gsl_interp_akima as *mut ffi::gsl_interp_type)
+        ffi_wrap!(gsl_interp_akima, gsl_interp_type)
     }
 
     /// Non-rounded Akima spline with periodic boundary conditions. This method uses the non-rounded corner algorithm of Wodicka.
     pub fn akima_periodic() -> InterpType {
-        ffi::FFI::wrap(ffi::gsl_interp_akima_periodic as *mut ffi::gsl_interp_type)
+        ffi_wrap!(gsl_interp_akima_periodic, gsl_interp_type)
     }
 }
 

--- a/src/types/mod.rs
+++ b/src/types/mod.rs
@@ -2,7 +2,14 @@
 // A rust binding for the GSL library by Guillaume Gomez (guillaume1.gomez@gmail.com)
 //
 
-pub use self::basis_spline::{BSpLineWorkspace, BSpLineDerivWorkspace};
+#[cfg(not(feature = "v2"))]
+pub use self::basis_spline::{
+    BSpLineWorkspace,
+    BSpLineDerivWorkspace
+};
+#[cfg(feature = "v2")]
+pub use self::basis_spline::BSpLineWorkspace;
+
 pub use self::chebyshev::ChebSeries;
 pub use self::combination::Combination;
 pub use self::complex::{ComplexF32, ComplexF64};

--- a/src/types/multifit_solver.rs
+++ b/src/types/multifit_solver.rs
@@ -240,6 +240,7 @@ pub struct MultiFitFunctionFdf<'r, T:'r> {
     pub params: &'r mut T
 }
 
+#[allow(dead_code)]
 struct LmderStateT {
     iter: usize,
     xnorm: f64,

--- a/src/types/ordinary_differential_equations.rs
+++ b/src/types/ordinary_differential_equations.rs
@@ -252,69 +252,87 @@ pub struct ODEiv2StepType {
 impl ODEiv2StepType {
     /// Explicit embedded Runge-Kutta (2, 3) method.
     pub fn rk2() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_rk2
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_rk2
+            }
         }
     }
 
     /// Explicit 4th order (classical) Runge-Kutta. Error estimation is carried out by the step doubling method. For more efficient
     /// estimate of the error, use the embedded methods described below.
     pub fn rk4() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_rk4
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_rk4
+            }
         }
     }
 
     /// Explicit embedded Runge-Kutta-Fehlberg (4, 5) method. This method is a good general-purpose integrator.
     pub fn rk45() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_rkf45
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_rkf45
+            }
         }
     }
 
     /// Explicit embedded Runge-Kutta Cash-Karp (4, 5) method.
     pub fn rkck() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_rkck
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_rkck
+            }
         }
     }
 
     /// Explicit embedded Runge-Kutta Prince-Dormand (8, 9) method.
     pub fn rk8pd() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_rk8pd
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_rk8pd
+            }
         }
     }
 
     /// Implicit Gaussian first order Runge-Kutta. Also known as implicit Euler or backward Euler method. Error estimation is carried out by
     /// the step doubling method. This algorithm requires the Jacobian and access to the driver object via gsl_odeiv2_step_set_driver.
     pub fn rk1imp() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_rk1imp
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_rk1imp
+            }
         }
     }
 
     /// Implicit Gaussian second order Runge-Kutta. Also known as implicit mid-point rule. Error estimation is carried out by the step doubling
     /// method. This stepper requires the Jacobian and access to the driver object via gsl_odeiv2_step_set_driver.
     pub fn rk2imp() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_rk2imp
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_rk2imp
+            }
         }
     }
 
     /// Implicit Gaussian 4th order Runge-Kutta. Error estimation is carried out by the step doubling method. This algorithm requires the
     /// Jacobian and access to the driver object via gsl_odeiv2_step_set_driver.
     pub fn rk4imp() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_rk4imp
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_rk4imp
+            }
         }
     }
 
     /// Implicit Bulirsch-Stoer method of Bader and Deuflhard. The method is generally suitable for stiff problems. This stepper requires
     /// the Jacobian.
     pub fn bsimp() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_bsimp
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_bsimp
+            }
         }
     }
 
@@ -322,8 +340,10 @@ impl ODEiv2StepType {
     /// implicit Adams-Moulton (corrector) methods in P(EC)^m functional iteration mode. Method order varies dynamically between 1 and 12.
     /// This stepper requires the access to the driver object via gsl_odeiv2_step_set_driver.
     pub fn msadams() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_msadams
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_msadams
+            }
         }
     }
 
@@ -332,8 +352,10 @@ impl ODEiv2StepType {
     /// non-linear equations. Method order varies dynamically between 1 and 5. The method is generally suitable for stiff problems. This
     /// stepper requires the Jacobian and the access to the driver object via gsl_odeiv2_step_set_driver.
     pub fn msbdf() -> ODEiv2StepType {
-        ODEiv2StepType {
-            t: ffi::gsl_odeiv2_step_msbdf
+        unsafe {
+            ODEiv2StepType {
+                t: ffi::gsl_odeiv2_step_msbdf
+            }
         }
     }
 }
@@ -517,14 +539,18 @@ pub struct ODEiv2ControlType {
 
 impl ODEiv2ControlType {
     pub fn scaled() -> ODEiv2ControlType {
-        ODEiv2ControlType {
-            t: ffi::gsl_odeiv2_control_scaled
+        unsafe {
+            ODEiv2ControlType {
+                t: ffi::gsl_odeiv2_control_scaled
+            }
         }
     }
 
     pub fn standard() -> ODEiv2ControlType {
-        ODEiv2ControlType {
-            t: ffi::gsl_odeiv2_control_standard
+        unsafe {
+            ODEiv2ControlType {
+                t: ffi::gsl_odeiv2_control_standard
+            }
         }
     }
 }

--- a/src/types/qrng.rs
+++ b/src/types/qrng.rs
@@ -125,30 +125,38 @@ impl QRngType {
     /// This generator uses the algorithm described in Bratley, Fox, Niederreiter, ACM Trans. Model. Comp. Sim. 2, 195 (1992). It is valid
     /// up to 12 dimensions.
     pub fn niederreiter_2() -> QRngType {
-        QRngType {
-            t: ffi::gsl_qrng_niederreiter_2
+        unsafe {
+            QRngType {
+                t: ffi::gsl_qrng_niederreiter_2
+            }
         }
     }
 
     /// This generator uses the Sobol sequence described in Antonov, Saleev, USSR Comput. Maths. Math. Phys. 19, 252 (1980). It is valid
     /// up to 40 dimensions.
     pub fn sobol() -> QRngType {
-        QRngType {
-            t: ffi::gsl_qrng_sobol
+        unsafe {
+            QRngType {
+                t: ffi::gsl_qrng_sobol
+            }
         }
     }
 
     /// These generators use the Halton and reverse Halton sequences described in J.H. Halton, Numerische Mathematik 2, 84-90 (1960) and
     /// B. Vandewoestyne and R. Cools Computational and Applied Mathematics 189, 1&2, 341-361 (2006). They are valid up to 1229 dimensions.
     pub fn halton() -> QRngType {
-        QRngType {
-            t: ffi::gsl_qrng_halton
+        unsafe {
+            QRngType {
+                t: ffi::gsl_qrng_halton
+            }
         }
     }
 
     pub fn reversehalton() -> QRngType {
-        QRngType {
-            t: ffi::gsl_qrng_reversehalton
+        unsafe {
+            QRngType {
+                t: ffi::gsl_qrng_reversehalton
+            }
         }
     }
 }

--- a/src/types/rng.rs
+++ b/src/types/rng.rs
@@ -199,7 +199,7 @@ impl Rng {
 
     /// Equivalent to DefaultRngSeed
     pub fn default_seed() -> usize {
-        ffi::gsl_rng_default_seed as usize
+        unsafe { ffi::gsl_rng_default_seed as usize }
     }
 }
 
@@ -339,7 +339,7 @@ impl ffi::FFI<ffi::gsl_rng_type> for RngType {
 }
 
 pub fn default() -> RngType {
-    ffi::FFI::wrap(ffi::gsl_rng_default as *mut ffi::gsl_rng_type)
+    ffi_wrap!(gsl_rng_default, gsl_rng_type)
 }
 
 /// The functions described above make no reference to the actual algorithm used. This is deliberate so that you can switch algorithms without having
@@ -367,7 +367,7 @@ pub mod algorithms {
     /// procedures could cause spurious artifacts for some seed values. They are still available through the alternative generators gsl_rng_mt19937_1999 and
     /// gsl_rng_mt19937_1998.
     pub fn mt19937() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_mt19937 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_mt19937, gsl_rng_type)
     }
 
     /// The generator ranlxs0 is a second-generation version of the RANLUX algorithm of Lüscher, which produces “luxury random numbers”. This generator
@@ -378,7 +378,7 @@ pub mod algorithms {
     /// 
     /// Note that the range of allowed seeds for this generator is [0,2^31-1]. Higher seed values are wrapped modulo 2^31.
     pub fn ranlxs0() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranlxs0 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranlxs0, gsl_rng_type)
     }
 
     /// The generator ranlxs0 is a second-generation version of the RANLUX algorithm of Lüscher, which produces “luxury random numbers”. This generator
@@ -389,7 +389,7 @@ pub mod algorithms {
     /// 
     /// Note that the range of allowed seeds for this generator is [0,2^31-1]. Higher seed values are wrapped modulo 2^31.
     pub fn ranlxs1() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranlxs1 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranlxs1, gsl_rng_type)
     }
 
     /// The generator ranlxs0 is a second-generation version of the RANLUX algorithm of Lüscher, which produces “luxury random numbers”. This generator
@@ -400,19 +400,19 @@ pub mod algorithms {
     /// 
     /// Note that the range of allowed seeds for this generator is [0,2^31-1]. Higher seed values are wrapped modulo 2^31.
     pub fn ranlxs2() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranlxs2 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranlxs2, gsl_rng_type)
     }
 
     /// This generator produces double precision output (48 bits) from the RANLXS generator. The library provides two luxury levels ranlxd1 and ranlxd2,
     /// in increasing order of strength.
     pub fn ranlxd1() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranlxd1 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranlxd1, gsl_rng_type)
     }
 
     /// This generator produces double precision output (48 bits) from the RANLXS generator. The library provides two luxury levels ranlxd1 and ranlxd2,
     /// in increasing order of strength.
     pub fn ranlxd2() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranlxd2 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranlxd2, gsl_rng_type)
     }
 
     /// The ranlux generator is an implementation of the original algorithm developed by Lüscher. It uses a lagged-fibonacci-with-skipping algorithm to
@@ -428,7 +428,7 @@ pub mod algorithms {
     /// M. Lüscher, “A portable high-quality random number generator for lattice field theory calculations”, Computer Physics Communications, 79 (1994) 100–110.
     /// F. James, “RANLUX: A Fortran implementation of the high-quality pseudo-random number generator of Lüscher”, Computer Physics Communications, 79 (1994) 111–114
     pub fn ranlux() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranlux as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranlux, gsl_rng_type)
     }
 
     /// The ranlux generator is an implementation of the original algorithm developed by Lüscher. It uses a lagged-fibonacci-with-skipping algorithm to
@@ -444,7 +444,7 @@ pub mod algorithms {
     /// M. Lüscher, “A portable high-quality random number generator for lattice field theory calculations”, Computer Physics Communications, 79 (1994) 100–110.
     /// F. James, “RANLUX: A Fortran implementation of the high-quality pseudo-random number generator of Lüscher”, Computer Physics Communications, 79 (1994) 111–114
     pub fn ranlux389() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranlux389 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranlux389, gsl_rng_type)
     }
 
     /// This is a combined multiple recursive generator by L’Ecuyer. Its sequence is,
@@ -462,7 +462,7 @@ pub mod algorithms {
     /// 
     /// P. L’Ecuyer, “Combined Multiple Recursive Random Number Generators”, Operations Research, 44, 5 (1996), 816–822.
     pub fn cmrg() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_cmrg as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_cmrg, gsl_rng_type)
     }
 
     /// This is a fifth-order multiple recursive generator by L’Ecuyer, Blouin and Coutre. Its sequence is,
@@ -475,7 +475,7 @@ pub mod algorithms {
     /// 
     /// P. L’Ecuyer, F. Blouin, and R. Coutre, “A search for good multiple recursive random number generators”, ACM Transactions on Modeling and Computer Simulation 3, 87–98 (1993).
     pub fn mrg() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_mrg as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_mrg, gsl_rng_type)
     }
 
     /// This is a maximally equidistributed combined Tausworthe generator by L’Ecuyer. The sequence is,
@@ -501,7 +501,7 @@ pub mod algorithms {
     /// 
     /// The generator gsl_rng_taus2 should now be used in preference to gsl_rng_taus.
     pub fn taus() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_taus as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_taus, gsl_rng_type)
     }
 
     /// This is a maximally equidistributed combined Tausworthe generator by L’Ecuyer. The sequence is,
@@ -527,7 +527,7 @@ pub mod algorithms {
     /// 
     /// The generator gsl_rng_taus2 should now be used in preference to gsl_rng_taus.
     pub fn taus2() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_taus2 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_taus2, gsl_rng_type)
     }
 
     /// The gfsr4 generator is like a lagged-fibonacci generator, and produces each number as an xor’d sum of four previous values.
@@ -554,7 +554,7 @@ pub mod algorithms {
     /// 
     /// Robert M. Ziff, “Four-tap shift-register-sequence random-number generators”, Computers in Physics, 12(4), Jul/Aug 1998, pp 385–392.
     pub fn gfsr4() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_gfsr4 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_gfsr4, gsl_rng_type)
     }
 }
 
@@ -574,7 +574,7 @@ pub mod unix {
     /// with a = 1103515245, c = 12345 and m = 2^31. The seed specifies the initial value, x_1. The period of this generator is 2^31, and it
     /// uses 1 word of storage per generator.
     pub fn rand() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_rand as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_rand, gsl_rng_type)
     }
 
     /// These generators implement the random family of functions, a set of linear feedback shift register generators originally used in BSD
@@ -596,7 +596,7 @@ pub mod unix {
     /// gsl_rng_random_bsd has been made equivalent to gsl_rng_random128_bsd. Corresponding versions of the libc5 and glibc2 generators are
     /// also available, with the names gsl_rng_random8_libc5, gsl_rng_random8_glibc2, etc.
     pub fn random_bsd() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_random_bsd as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_random_bsd, gsl_rng_type)
     }
 
     /// These generators implement the random family of functions, a set of linear feedback shift register generators originally used in BSD
@@ -618,7 +618,7 @@ pub mod unix {
     /// gsl_rng_random_bsd has been made equivalent to gsl_rng_random128_bsd. Corresponding versions of the libc5 and glibc2 generators are
     /// also available, with the names gsl_rng_random8_libc5, gsl_rng_random8_glibc2, etc.
     pub fn random_libc5() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_random_libc5 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_random_libc5, gsl_rng_type)
     }
 
     /// These generators implement the random family of functions, a set of linear feedback shift register generators originally used in BSD
@@ -640,7 +640,7 @@ pub mod unix {
     /// gsl_rng_random_bsd has been made equivalent to gsl_rng_random128_bsd. Corresponding versions of the libc5 and glibc2 generators are
     /// also available, with the names gsl_rng_random8_libc5, gsl_rng_random8_glibc2, etc.
     pub fn random_glic2() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_random_glibc2 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_random_glibc2, gsl_rng_type)
     }
 
     /// This is the Unix rand48 generator. Its sequence is
@@ -653,7 +653,7 @@ pub mod unix {
     /// is equivalent to the function drand48. Note that some versions of the GNU C Library contained a bug in mrand48 function which caused
     /// it to produce different results (only the lower 16-bits of the return value were set).
     pub fn rand48() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_rand48 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_rand48, gsl_rng_type)
     }
 }
 
@@ -686,13 +686,13 @@ pub mod other {
     /// 
     /// The period of this generator is 2^46.
     pub fn ranf() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranf as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranf, gsl_rng_type)
     }
 
     /// This is the RANMAR lagged-fibonacci generator of Marsaglia, Zaman and Tsang. It is a 24-bit generator, originally designed for single-precision IEEE floating point numbers. 
     /// It was included in the CERNLIB high-energy physics library.
     pub fn ranmar() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_ranmar as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_ranmar, gsl_rng_type)
     }
 
     /// This is the shift-register generator of Kirkpatrick and Stoll. The sequence is based on the recurrence
@@ -704,7 +704,7 @@ pub mod other {
     /// 
     /// S. Kirkpatrick and E. Stoll, “A very fast shift-register sequence random number generator”, Journal of Computational Physics, 40, 517–526 (1981)
     pub fn r250() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_r250 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_r250, gsl_rng_type)
     }
 
     /// This is an earlier version of the twisted generalized feedback shift-register generator, and has been superseded by the development of MT19937. However, it is
@@ -714,7 +714,7 @@ pub mod other {
     /// 
     /// Makoto Matsumoto and Yoshiharu Kurita, “Twisted GFSR Generators II”, ACM Transactions on Modelling and Computer Simulation, Vol. 4, No. 3, 1994, pages 254–266.
     pub fn tt800() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_tt800 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_tt800, gsl_rng_type)
     }
 
     /// This is the VAX generator MTH$RANDOM. Its sequence is,
@@ -723,7 +723,7 @@ pub mod other {
     /// 
     /// with a = 69069, c = 1 and m = 2^32. The seed specifies the initial value, x_1. The period of this generator is 2^32 and it uses 1 word of storage per generator.
     pub fn vax() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_vax as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_vax, gsl_rng_type)
     }
 
     /// This is the random number generator from the INMOS Transputer Development system. Its sequence is,
@@ -732,7 +732,7 @@ pub mod other {
     /// 
     /// with a = 1664525 and m = 2^32. The seed specifies the initial value, x_1.
     pub fn transputer() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_transputer as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_transputer, gsl_rng_type)
     }
 
     /// This is the IBM RANDU generator. Its sequence is
@@ -741,7 +741,7 @@ pub mod other {
     /// 
     /// with a = 65539 and m = 2^31. The seed specifies the initial value, x_1. The period of this generator was only 2^29. It has become a textbook example of a poor generator.
     pub fn randu() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_randu as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_randu, gsl_rng_type)
     }
 
     /// This is Park and Miller’s “minimal standard” MINSTD generator, a simple linear congruence which takes care to avoid the major pitfalls of such algorithms. Its sequence is,
@@ -756,24 +756,24 @@ pub mod other {
     /// 
     /// Park and Miller, "Random Number Generators: Good ones are hard to find", Communications of the ACM, October 1988, Volume 31, No 10, pages 1192–1201.
     pub fn minstd() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_minstd as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_minstd, gsl_rng_type)
     }
 
     /// This is a reimplementation of the 16-bit SLATEC random number generator RUNIF. A generalization of the generator to 32 bits is provided by gsl_rng_uni32.
     /// The original source code is available from NETLIB.
     pub fn uni() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_uni as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_uni, gsl_rng_type)
     }
 
     /// This is a reimplementation of the 16-bit SLATEC random number generator RUNIF. A generalization of the generator to 32 bits is provided by gsl_rng_uni32.
     /// The original source code is available from NETLIB.
     pub fn uni32() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_uni32 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_uni32, gsl_rng_type)
     }
 
     /// This is the SLATEC random number generator RAND. It is ancient. The original source code is available from NETLIB.
     pub fn slatec() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_slatec as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_slatec, gsl_rng_type)
     }
 
     /// This is the ZUFALL lagged Fibonacci series generator of Peterson. Its sequence is,
@@ -785,7 +785,7 @@ pub mod other {
     /// 
     /// W. Petersen, “Lagged Fibonacci Random Number Generators for the NEC SX-3”, International Journal of High Speed Computing (1994).
     pub fn zuf() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_zuf as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_zuf, gsl_rng_type)
     }
 
     /// This is a second-order multiple recursive generator described by Knuth in Seminumerical Algorithms, 3rd Ed., page 108. Its sequence is,
@@ -794,21 +794,21 @@ pub mod other {
     /// 
     /// with a_1 = 271828183, a_2 = 314159269, and m = 2^31 - 1.
     pub fn knuthran2() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_knuthran2 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_knuthran2, gsl_rng_type)
     }
 
     /// This is a second-order multiple recursive generator described by Knuth in Seminumerical Algorithms, 3rd Ed., Section 3.6. Knuth provides
     /// its C code. The updated routine gsl_rng_knuthran2002 is from the revised 9th printing and corrects some weaknesses in the earlier version,
     /// which is implemented as gsl_rng_knuthran.
     pub fn knuthran2002() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_knuthran2002 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_knuthran2002, gsl_rng_type)
     }
 
     /// This is a second-order multiple recursive generator described by Knuth in Seminumerical Algorithms, 3rd Ed., Section 3.6. Knuth provides
     /// its C code. The updated routine gsl_rng_knuthran2002 is from the revised 9th printing and corrects some weaknesses in the earlier version,
     /// which is implemented as gsl_rng_knuthran.
     pub fn knuthran() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_knuthran as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_knuthran, gsl_rng_type)
     }
 
     /// This multiplicative generator is taken from Knuth’s Seminumerical Algorithms, 3rd Ed., pages 106–108. Their sequence is,
@@ -819,7 +819,7 @@ pub mod other {
     /// m = 2^32, Fishman18: a = 62089911, m = 2^31 - 1, Fishman20: a = 48271, m = 2^31 - 1, L’Ecuyer: a = 40692, m = 2^31 - 249,
     /// Waterman: a = 1566083941, m = 2^32.
     pub fn borosh13() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_borosh13 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_borosh13, gsl_rng_type)
     }
 
     /// This multiplicative generator is taken from Knuth’s Seminumerical Algorithms, 3rd Ed., pages 106–108. Their sequence is,
@@ -830,7 +830,7 @@ pub mod other {
     /// m = 2^32, Fishman18: a = 62089911, m = 2^31 - 1, Fishman20: a = 48271, m = 2^31 - 1, L’Ecuyer: a = 40692, m = 2^31 - 249,
     /// Waterman: a = 1566083941, m = 2^32.
     pub fn fishman18() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_fishman18 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_fishman18, gsl_rng_type)
     }
 
     /// This multiplicative generator is taken from Knuth’s Seminumerical Algorithms, 3rd Ed., pages 106–108. Their sequence is,
@@ -841,7 +841,7 @@ pub mod other {
     /// m = 2^32, Fishman18: a = 62089911, m = 2^31 - 1, Fishman20: a = 48271, m = 2^31 - 1, L’Ecuyer: a = 40692, m = 2^31 - 249,
     /// Waterman: a = 1566083941, m = 2^32.
     pub fn fishman20() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_fishman20 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_fishman20, gsl_rng_type)
     }
 
     /// This multiplicative generator is taken from Knuth’s Seminumerical Algorithms, 3rd Ed., pages 106–108. Their sequence is,
@@ -852,7 +852,7 @@ pub mod other {
     /// m = 2^32, Fishman18: a = 62089911, m = 2^31 - 1, Fishman20: a = 48271, m = 2^31 - 1, L’Ecuyer: a = 40692, m = 2^31 - 249,
     /// Waterman: a = 1566083941, m = 2^32.
     pub fn lecuyer21() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_lecuyer21 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_lecuyer21, gsl_rng_type)
     }
 
     /// This multiplicative generator is taken from Knuth’s Seminumerical Algorithms, 3rd Ed., pages 106–108. Their sequence is,
@@ -863,7 +863,7 @@ pub mod other {
     /// m = 2^32, Fishman18: a = 62089911, m = 2^31 - 1, Fishman20: a = 48271, m = 2^31 - 1, L’Ecuyer: a = 40692, m = 2^31 - 249,
     /// Waterman: a = 1566083941, m = 2^32.
     pub fn waterman14() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_waterman14 as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_waterman14, gsl_rng_type)
     }
 
     /// This is the L’Ecuyer–Fishman random number generator. It is taken from Knuth’s Seminumerical Algorithms, 3rd Ed., page 108. Its sequence is,
@@ -872,7 +872,7 @@ pub mod other {
     /// 
     /// with m = 2^31 - 1. x_n and y_n are given by the fishman20 and lecuyer21 algorithms. The seed specifies the initial value, x_1.
     pub fn fishman2x() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_fishman2x as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_fishman2x, gsl_rng_type)
     }
 
     /// This is the Coveyou random number generator. It is taken from Knuth’s Seminumerical Algorithms, 3rd Ed., Section 3.2.2. Its sequence is,
@@ -881,6 +881,6 @@ pub mod other {
     /// 
     /// with m = 2^32. The seed specifies the initial value, x_1.
     pub fn coveyou() -> RngType {
-        ffi::FFI::wrap(ffi::gsl_rng_coveyou as *mut ffi::gsl_rng_type)
+        ffi_wrap!(gsl_rng_coveyou, gsl_rng_type)
     }
 }

--- a/src/types/wavelet_transforms.rs
+++ b/src/types/wavelet_transforms.rs
@@ -116,46 +116,58 @@ impl WaveletType {
     /// This is the Daubechies wavelet family of maximum phase with k/2 vanishing moments. The implemented wavelets are k=4, 6, …, 20, with
     /// k even.
     pub fn daubechies() -> WaveletType {
-        WaveletType {
-            t: ffi::gsl_wavelet_daubechies
+        unsafe {
+            WaveletType {
+                t: ffi::gsl_wavelet_daubechies
+            }
         }
     }
 
     /// This is the Daubechies wavelet family of maximum phase with k/2 vanishing moments. The implemented wavelets are k=4, 6, …, 20, with
     /// k even.
     pub fn daubechies_centered() -> WaveletType {
-        WaveletType {
-            t: ffi::gsl_wavelet_daubechies_centered
+        unsafe {
+            WaveletType {
+                t: ffi::gsl_wavelet_daubechies_centered
+            }
         }
     }
 
     /// This is the Haar wavelet. The only valid choice of k for the Haar wavelet is k=2.
     pub fn haar() -> WaveletType {
-        WaveletType {
-            t: ffi::gsl_wavelet_haar
+        unsafe {
+            WaveletType {
+                t: ffi::gsl_wavelet_haar
+            }
         }
     }
 
     /// This is the Haar wavelet. The only valid choice of k for the Haar wavelet is k=2.
     pub fn haar_centered() -> WaveletType {
-        WaveletType {
-            t: ffi::gsl_wavelet_haar_centered
+        unsafe {
+            WaveletType {
+                t: ffi::gsl_wavelet_haar_centered
+            }
         }
     }
 
     /// This is the biorthogonal B-spline wavelet family of order (i,j). The implemented values of k = 100*i + j are 103, 105, 202, 204,
     /// 206, 208, 301, 303, 305 307, 309.
     pub fn bspline() -> WaveletType {
-        WaveletType {
-            t: ffi::gsl_wavelet_bspline
+        unsafe {
+            WaveletType {
+                t: ffi::gsl_wavelet_bspline
+            }
         }
     }
 
     /// This is the biorthogonal B-spline wavelet family of order (i,j). The implemented values of k = 100*i + j are 103, 105, 202, 204,
     /// 206, 208, 301, 303, 305 307, 309.
     pub fn bspline_centered() -> WaveletType {
-        WaveletType {
-            t: ffi::gsl_wavelet_bspline_centered
+        unsafe {
+            WaveletType {
+                t: ffi::gsl_wavelet_bspline_centered
+            }
         }
     }
 }

